### PR TITLE
Performance improvements

### LINKED ISSUE #15: Improve performance and memory footprint
There's a few things we can do to improve the two - remove offline merging since our online algo is good enough, discard received stream data once we no longer need it.

### PR DESCRIPTION
**Pull request by Wesmania**
Fixes #15.

### CODE REVIEW

File: replayserver/streams/base.py [hunk 1]
```diff
@@ -154,22 +173,76 @@ class ConcreteDataMixin:
...
+        if isinstance(s, slice):
+            return self._get_slice(s)
+        if s >= 0:
+            if s < self._discarded_data:
+                raise IndexError
+            s -= self._discarded_data
         return self._data[s]
 
+    def _get_slice(self, s):
+        s, e, st = s.start, s.stop, s.step
```
Comment (norraxx) [inline replayserver/streams/base.py]: `M != a`

`!` = do not
`M` = I
`a` = anagram
`=` = like

File: replayserver/streams/delayed.py [hunk 2]
```diff
@@ -66,17 +66,41 @@ def _data_length(self):
         return min(len(self._stream.data), self._current_position)
 
     def _data_slice(self, s):
+        if isinstance(s, slice):
+            return self._get_slice(s)
```
Comment (norraxx) [inline replayserver/streams/delayed.py]: if s is instance of global method method `slice` https://www.programiz.com/python-programming/methods/built-in/slice
call `_get_slice` with `s`, where on function `s` will be variable `stop`...

File: replayserver/streams/delayed.py [hunk 3]
```diff
@@ -66,17 +66,41 @@ def _data_length(self):
         return min(len(self._stream.data), self._current_position)
 
     def _data_slice(self, s):
```
Comment (norraxx) [inline replayserver/streams/delayed.py]: without comment, typing and without proper naming this code is available only for one person.

File: replayserver/receive/memprefix.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        else:
+            end = start + chunk
+
+    for c1, c2 in zip(b1[start:end], b2[start:end]):
+        if c1 == c2:
+            start += 1
+        else:
+            break
+
+    return start
```
[the remark this answers is not in the document]
Comment (Wesmania) [reply replayserver/receive/memprefix.py]: Function is documented and algorithm is trivial enough to not need comments

File: replayserver/streams/base.py (hunk 1 above)
Comment (Wesmania) [reply replayserver/streams/base.py]: not going to use verbose names for mucking around with arithmetics, shorthands are obvious

File: replayserver/streams/delayed.py (hunk 2 above)
Comment (Wesmania) [reply replayserver/streams/delayed.py]: `slice is type(slice(1,1,1)) == True`

File: replayserver/streams/delayed.py (hunk 3 above)
Comment (Wesmania) [reply replayserver/streams/delayed.py]: interface documentation is in replayserver/streams/base.py

File: replayserver/streams/base.py (hunk 1 above)
Comment (Wesmania) [reply replayserver/streams/base.py]: There's actually an `indices` method of `slice` that avoids this mucking about. Good to know, glad this code was revisited.

File: replayserver/streams/delayed.py (hunk 2 above)
Comment (norraxx) [reply replayserver/streams/delayed.py]: ??

Comment (Wesmania) [reply replayserver/streams/delayed.py]: `slice` is a class, not a function

File: replayserver/receive/memprefix.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Brutus5000) [reply replayserver/receive/memprefix.py]: What is b1 and b2? No type and the docs calls it b1 and b2. This is not helpful.

File: replayserver/streams/delayed.py (hunk 2 above)
Comment (norraxx) [reply replayserver/streams/delayed.py]: Classes in python should be `CamelCase`

Comment (norraxx) [reply replayserver/streams/delayed.py]: When `slice` is internal python function. from __builtins__ import slice

Comment (Wesmania) [reply replayserver/streams/delayed.py]: Straight from the interpreter:
```
>>> slice
<class 'slice'>
```

**Pull request merged**